Add a K8s Job to configure a default domain.

tcnghia opened the pull request:
This adds a simple Kubernetes Job set up a default DNS entry based on `xip.io` with the IP address of the Istio Ingress Gateway.

<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Fixes #2968 

## Proposed Changes

*  Add a simple Kubernetes Job set up a default DNS entry based on `xip.io` with the IP address of the Istio Ingress Gateway.  The job does nothing when seeing IngressGateway that has domains instead of IPs.
*  Make e2e tests use `--resolvabledomain=true` when `istio-ingressgateway` has IP, except for base version of upgrade tests, which isn't yet possible.


**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
* Add a simple post-install Kubernetes Job set up a default DNS entry based on `xip.io` with the IP address of the Istio Ingress Gateway.  The job does nothing when seeing IngressGateway that has domains instead of IPs.  This job only runs once to give a better default than `example.com` when starting up.
```

tcnghia commented on cmd/default-domain/main.go:
```diff
@@ -0,0 +1,167 @@
+	if hostname != nil {
+		return hostname, nil
+	}
+	return nil, errors.New("Istio gateway does not have an assigned external address.")
```
DOne

tcnghia commented:
/assign @vagababov

tcnghia commented:
Rerun tests

/test pull-knative-serving-integration-tests
/test pull-knative-serving-upgrade-tests

vagababov commented on cmd/default-domain/main.go:
```diff
@@ -0,0 +1,181 @@
+
+func waitForConfigMap(kubeClient *kubernetes.Clientset, name string) (*corev1.ConfigMap, error) {
+	logger := logging.FromContext(context.TODO()).Named(appName)
+	for {
```
use k8s.wait?

tcnghia replied:
Done

ZhiminXiang commented on cmd/default-domain/main.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+func waitForConfigMap(kubeClient *kubernetes.Clientset, name string) (*corev1.ConfigMap, error) {
+	logger := logging.FromContext(context.TODO()).Named(appName)
```
Seems like context.Background() is preferable. See https://github.com/knative/serving/pull/3476

tcnghia replied:
Done

vagababov commented on cmd/default-domain/main.go:
```diff
@@ -0,0 +1,181 @@
+	var ip *corev1.LoadBalancerIngress
+	var hostname *corev1.LoadBalancerIngress
+	for i, ing := range svc.Status.LoadBalancer.Ingress {
+		if ing.IP != "" {
```
The code looks for last, not first.

tcnghia replied:
Updated

vagababov commented on cmd/default-domain/main.go:
```diff
@@ -0,0 +1,181 @@
+
+func waitForIngressGatewayAddress(kubeclient *kubernetes.Clientset) *corev1.LoadBalancerIngress {
+	logger := logging.FromContext(context.TODO()).Named(appName)
+	for {
```
Ditto about k8s wait library.

tcnghia replied:
Done

vagababov commented on cmd/default-domain/main.go:
```diff
@@ -0,0 +1,181 @@
+	// and send it back to the API server.
+	domain := fmt.Sprintf("%s.%s", address.IP, *magicDNS)
+	domainCM.Data[domain] = ""
+	_, err = kubeClient.CoreV1().ConfigMaps(system.Namespace()).Update(domainCM)
```
Join these two lines.

tcnghia replied:
Done

vagababov commented on test/crd.go:
```diff
@@ -259,9 +259,16 @@ func SubServiceNameForTest(t *testing.T, subsvc string) string {
+	// First, replace things like "GRPC" to "Grpc" to be one camel-case token.
+	// This makes sure we turn GRPC into "grpc" and not "g-r-p-c".
+	sanitized := s
+	for orig, new := range map[string]string{"GRPC": "Grpc"} {
```
well, I can think of HTTP, H2C :)

tcnghia replied:
Yeah, will split to a different PR.

vagababov commented on cmd/default-domain/main.go:
```diff
@@ -159,7 +161,10 @@ func main() {
-	address := waitForIngressGatewayAddress(kubeClient)
+	address, err := waitForIngressGatewayAddress(kubeClient)
+	if err != nil {
+		logger.Fatalw("Error waiting for IngressGateway address")
```
Add the error or remove "w".

tcnghia replied:
Done

tcnghia commented:
/assign @mattmoor 
for OWNERS

mattmoor commented on config/204-istio-default-domain.yaml:
```diff
@@ -0,0 +1,43 @@
+kind: Job
+metadata:
+  generateName: default-domain
+  name: default-domain
```
You should not have both of these.

tcnghia replied:
Done

mattmoor commented on config/204-istio-default-domain.yaml:
```diff
@@ -0,0 +1,43 @@
+  name: default-domain
+  namespace: knative-serving
+  labels:
+    app: "default-domain"
```
Perhaps we should include labels on istio things so folks can install without istio via `kubectl apply -f release.yaml -l foo!=istio` 🤔

tcnghia replied:
Done

mattmoor commented on cmd/default-domain/main.go:
```diff
@@ -0,0 +1,185 @@
+
+func waitForConfigMap(kubeClient *kubernetes.Clientset, name string) (cm *corev1.ConfigMap, waitErr error) {
+	logger := logging.FromContext(context.Background()).Named(appName)
+	waitErr = wait.PollImmediate(pollInterval, waitTimeout, func() (done bool, err error) {
```
I don't understand why this would use the API vs. simply volume mounting it?

mattmoor replied:
Then we eliminate the wait because the pod won't start until it exists

tcnghia replied:
Done

mattmoor commented on test/e2e/grpc_test.go:
```diff
@@ -35,14 +35,31 @@ import (
-	t.Logf("Connecting to grpc-ping using host %q and authority %q", host, domain)
-	conn, err := grpc.Dial(
+func dial(host, domain string) (*grpc.ClientConn, error) {
+	if host != domain {
```
when does this happen?

tcnghia replied:
This is when using ResolvableDomain = false, see line 186

mattmoor commented on test/e2e-upgrade-tests.sh:
```diff
@@ -70,16 +70,22 @@ initialize $@
 header "Running postdowngrade tests"
-go_test_e2e -tags=postdowngrade -timeout=${TIMEOUT} ./test/upgrade || fail_test
+go_test_e2e -tags=postdowngrade -timeout=${TIMEOUT} ./test/upgrade \
+  --resolvabledomain=$(use_resolvable_domain) || fail_test
```
This works because we've run the Job?

tcnghia replied:
I'm moving to test/config so all 3 times are using the same now.

tcnghia commented on config/post-install/100-istio-default-domain.yaml:
```diff
@@ -0,0 +1,57 @@
+  namespace: knative-serving
+  labels:
+    app: "default-domain"
+    networking.knative.dev/provider: "istio"
```
@mattmoor let me know if you would prefer a different label here.

mattmoor replied:
This was mainly relevant in the context of the other file renames that you reverted.  I'd label all of the istio stuff holistically as we plan to split off that implementation.

tcnghia commented:
@mattmoor RFAL.  Things I've changed:

* Revert the test/e2e_flags.go defaulting, since may be not everyone dev workflow allows that.
* Move template to `config/post-install` and symlinked.

and addressed all the PR feedbacks.

tcnghia commented:
/test pull-knative-serving-integration-tests

ZhiminXiang commented on cmd/default-domain/main.go:
```diff
@@ -0,0 +1,175 @@
+	}
+	// Parse the service name to determine which Kubernetes Service resource to fetch.
+	serviceURL := istioConfig.IngressGateways[0].ServiceURL
+	parts := strings.SplitN(serviceURL, ".", 3)
```
Could you please add a comment for the format `serviceURL`? It would make it easy to understand why it is `3` here.

tcnghia replied:
Done

ZhiminXiang commented:
/lgtm

bbrowning commented:
This is something I think we should move into an Operator instead of a post-install Kubernetes Job longer-term. cc @jcrossley3

tcnghia commented:
@bbrowning +1 .  A post install job can not be reapplied  so won't be good.  I  removed it from `config/` into `config/post-install` for that reason.
